File: LyricFeverTests/MetadataMatcherTests.swift
import XCTest

final class MetadataMatcherTests: XCTestCase {
    func testTitleCandidatesIncludeBracketFreeVersion() {
        XCTAssertEqual(
            MetadataMatcher.titleCandidates(for: "Song Name (Live Version)"),
            ["Song Name (Live Version)", "Song Name"]
        )
    }

    func testExactTitleAndArtistRankHighest() {
        let exact = result(title: "Café", artist: "Artist")
        let partial = result(title: "Café (Live)", artist: "Artist")
        let unrelated = result(title: "Another Song", artist: "Someone Else")

        let sorted = MetadataMatcher.filteredAndSorted(
            [unrelated, partial, exact],
            trackName: "Cafe",
            artistName: "Artist"
        )

        XCTAssertEqual(sorted.map(\.songName), ["Café", "Café (Live)"])
    }

    func testDuplicateProviderResultIsRemoved() {
        let first = result(title: "Song", artist: "Artist")
        let duplicate = result(title: "song", artist: "artist")
        let sorted = MetadataMatcher.filteredAndSorted(
            [first, duplicate],
            trackName: "Song",
            artistName: "Artist"
        )
        XCTAssertEqual(sorted.count, 1)
    }

    func testPlausibleMatchRejectsShortUnrelatedCandidate() {
        XCTAssertFalse(MetadataMatcher.plausiblyMatches("A Very Long Song Name", "Song"))
        XCTAssertTrue(MetadataMatcher.plausiblyMatches("Song Name Remaster", "Song Name Remaster 2024"))
    }

    private func result(title: String, artist: String) -> SongResult {
        SongResult(
            lyricType: "Test",
            songName: title,
            albumName: "Album",
            artistName: artist,
            lyrics: [LyricLine(startTime: 0, words: "Line")]
        )
    }
}

File: Windows/tests/LyricFever.Core.Tests/MetadataMatcherTests.cs
using LyricFever.Core.Lyrics;
using Xunit;

namespace LyricFever.Core.Tests;

/// <summary>对照 macOS MetadataMatcherTests 移植。</summary>
public class MetadataMatcherTests
{
    [Fact]
    public void TitleCandidatesIncludeBracketFreeVersion()
    {
        Assert.Equal(
            new[] { "Song Name (Live Version)", "Song Name" },
            MetadataMatcher.TitleCandidates("Song Name (Live Version)"));
    }

    [Fact]
    public void ExactTitleAndArtistRankHighest()
    {
        var exact = Result("Café", "Artist");
        var partial = Result("Café (Live)", "Artist");
        var unrelated = Result("Another Song", "Someone Else");

        var sorted = MetadataMatcher.FilteredAndSorted(
            new List<SongResult> { unrelated, partial, exact },
            "Cafe", "Artist");

        Assert.Equal(new[] { "Café", "Café (Live)" }, sorted.Select(r => r.SongName));
    }

    [Fact]
    public void DuplicateProviderResultIsRemoved()
    {
        var first = Result("Song", "Artist");
        var duplicate = Result("song", "artist");
        var sorted = MetadataMatcher.FilteredAndSorted(
            new List<SongResult> { first, duplicate },
            "Song", "Artist");
        Assert.Single(sorted);
    }

    [Fact]
    public void PlausibleMatchRejectsShortUnrelatedCandidate()
    {
        Assert.False(MetadataMatcher.PlausiblyMatches("A Very Long Song Name", "Song"));
        Assert.True(MetadataMatcher.PlausiblyMatches("Song Name Remaster", "Song Name Remaster 2024"));
    }

    private static SongResult Result(string title, string artist) => new()
    {
        LyricType = "Test",
        SongName = title,
        AlbumName = "Album",
        ArtistName = artist,
        Lyrics = new List<LyricLine> { new(0, "Line") }
    };
}
